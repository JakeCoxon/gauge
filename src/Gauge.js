import React, { Component } from 'react';

import { describeArc, polarToCartesian, lerp } from './util.js'

const width = 500;
const height = 300;
const centerX = width / 2;
const centerY = height * 2/3;
const arcRadius = 130;
const lineRadius = arcRadius + 20;
const labelRadius = lineRadius + 20;
const minMaxLabelRadius = lineRadius + 10;
const minLabelPosition = polarToCartesian(centerX, centerY, minMaxLabelRadius, -90);
const maxLabelPosition = polarToCartesian(centerX, centerY, minMaxLabelRadius, 90);

export default function Gauge({ min, max, value, format }) {
  const clampedValue = Math.min(Math.max(value, min), max)
  const t = (clampedValue - min) / (max - min);
  const endLine = polarToCartesian(centerX, centerY, lineRadius, lerp(t, -90, 90))
  const endText = polarToCartesian(centerX, centerY, labelRadius, lerp(t, -90, 90))

  return <svg style={{ width, height }} viewBox={`0 0 ${width} ${height}`}>
    <path {...Gauge.attributes.arc} d={describeArc({ x: centerX, y: centerY + 1, radius: arcRadius, startAngle: 270, endAngle: 90 })} />
    <line x1={centerX} y1={centerY} x2={endLine.x} y2={endLine.y} stroke="black" />
    <text {...Gauge.attributes.textLabel} {...polarToCartesian(centerX, centerY, labelRadius, lerp(t, -90, 90))}>{format(value)}</text>
    <text {...Gauge.attributes.textLabel} {...minLabelPosition}>{format(min)}</text>
    <text {...Gauge.attributes.textLabel} {...maxLabelPosition}>{format(max)}</text>
  </svg>
}
Gauge.attributes = {
  textLabel: {
    fill: "rgba(0,0,0,0.7)",
    style: { fontFamily: 'Roboto, sans-serif' },
    textAnchor: "middle"
  },
  arc: {
    fill: "rgba(0,0,0,0.35)",
    stroke: "rgba(0,0,0,0.8)",
    strokeWidth: "4"
  }
}


