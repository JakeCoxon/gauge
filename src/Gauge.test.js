import React from 'react';
import ReactDOM from 'react-dom';
import Gauge from './Gauge.js';
import renderer from 'react-test-renderer';


const defaultFormat = x => x
const moneyFormat = x => `Â£${Math.floor(x)}`

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Gauge min={0} max={100} value={30} format={defaultFormat} />, div);
  ReactDOM.unmountComponentAtNode(div);
});


const testProps = [
  { min: 0, max: 100, value: 0, format: defaultFormat },
  { min: 0, max: 100, value: 30, format: defaultFormat },
  { min: 100, max: 200, value: 30, format: defaultFormat },
  { min: 0, max: 200, value: 300, format: defaultFormat },
  { min: 0, max: 200, value: 10, format: moneyFormat },
]

testProps.forEach(props => {
  it('renders props', () => {
    const tree = renderer
      .create(<Gauge {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
})


