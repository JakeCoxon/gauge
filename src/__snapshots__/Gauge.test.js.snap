// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders props 1`] = `
<svg
  style={
    Object {
      "height": 300,
      "width": 500,
    }
  }
  viewBox="0 0 500 300"
>
  <path
    d="M 380 201 A 130 130 0 0 0 120 201.00000000000003 "
    fill="rgba(0,0,0,0.35)"
    stroke="rgba(0,0,0,0.8)"
    strokeWidth="4"
  />
  <line
    stroke="black"
    x1={250}
    x2={100}
    y1={200}
    y2={199.99999999999997}
  />
  <text
    fill="rgba(0,0,0,0.7)"
    style={
      Object {
        "fontFamily": "Roboto, sans-serif",
      }
    }
    textAnchor="middle"
    x={80}
    y={199.99999999999997}
  >
    0
  </text>
  <text
    fill="rgba(0,0,0,0.7)"
    style={
      Object {
        "fontFamily": "Roboto, sans-serif",
      }
    }
    textAnchor="middle"
    x={90}
    y={199.99999999999997}
  >
    0
  </text>
  <text
    fill="rgba(0,0,0,0.7)"
    style={
      Object {
        "fontFamily": "Roboto, sans-serif",
      }
    }
    textAnchor="middle"
    x={410}
    y={200}
  >
    100
  </text>
</svg>
`;

exports[`renders props 2`] = `
<svg
  style={
    Object {
      "height": 300,
      "width": 500,
    }
  }
  viewBox="0 0 500 300"
>
  <path
    d="M 380 201 A 130 130 0 0 0 120 201.00000000000003 "
    fill="rgba(0,0,0,0.35)"
    stroke="rgba(0,0,0,0.8)"
    strokeWidth="4"
  />
  <line
    stroke="black"
    x1={250}
    x2={161.83221215612906}
    y1={200}
    y2={78.64745084375788}
  />
  <text
    fill="rgba(0,0,0,0.7)"
    style={
      Object {
        "fontFamily": "Roboto, sans-serif",
      }
    }
    textAnchor="middle"
    x={150.07650711027958}
    y={62.46711095625892}
  >
    30
  </text>
  <text
    fill="rgba(0,0,0,0.7)"
    style={
      Object {
        "fontFamily": "Roboto, sans-serif",
      }
    }
    textAnchor="middle"
    x={90}
    y={199.99999999999997}
  >
    0
  </text>
  <text
    fill="rgba(0,0,0,0.7)"
    style={
      Object {
        "fontFamily": "Roboto, sans-serif",
      }
    }
    textAnchor="middle"
    x={410}
    y={200}
  >
    100
  </text>
</svg>
`;

exports[`renders props 3`] = `
<svg
  style={
    Object {
      "height": 300,
      "width": 500,
    }
  }
  viewBox="0 0 500 300"
>
  <path
    d="M 380 201 A 130 130 0 0 0 120 201.00000000000003 "
    fill="rgba(0,0,0,0.35)"
    stroke="rgba(0,0,0,0.8)"
    strokeWidth="4"
  />
  <line
    stroke="black"
    x1={250}
    x2={100}
    y1={200}
    y2={199.99999999999997}
  />
  <text
    fill="rgba(0,0,0,0.7)"
    style={
      Object {
        "fontFamily": "Roboto, sans-serif",
      }
    }
    textAnchor="middle"
    x={80}
    y={199.99999999999997}
  >
    30
  </text>
  <text
    fill="rgba(0,0,0,0.7)"
    style={
      Object {
        "fontFamily": "Roboto, sans-serif",
      }
    }
    textAnchor="middle"
    x={90}
    y={199.99999999999997}
  >
    100
  </text>
  <text
    fill="rgba(0,0,0,0.7)"
    style={
      Object {
        "fontFamily": "Roboto, sans-serif",
      }
    }
    textAnchor="middle"
    x={410}
    y={200}
  >
    200
  </text>
</svg>
`;

exports[`renders props 4`] = `
<svg
  style={
    Object {
      "height": 300,
      "width": 500,
    }
  }
  viewBox="0 0 500 300"
>
  <path
    d="M 380 201 A 130 130 0 0 0 120 201.00000000000003 "
    fill="rgba(0,0,0,0.35)"
    stroke="rgba(0,0,0,0.8)"
    strokeWidth="4"
  />
  <line
    stroke="black"
    x1={250}
    x2={400}
    y1={200}
    y2={200}
  />
  <text
    fill="rgba(0,0,0,0.7)"
    style={
      Object {
        "fontFamily": "Roboto, sans-serif",
      }
    }
    textAnchor="middle"
    x={420}
    y={200}
  >
    300
  </text>
  <text
    fill="rgba(0,0,0,0.7)"
    style={
      Object {
        "fontFamily": "Roboto, sans-serif",
      }
    }
    textAnchor="middle"
    x={90}
    y={199.99999999999997}
  >
    0
  </text>
  <text
    fill="rgba(0,0,0,0.7)"
    style={
      Object {
        "fontFamily": "Roboto, sans-serif",
      }
    }
    textAnchor="middle"
    x={410}
    y={200}
  >
    200
  </text>
</svg>
`;

exports[`renders props 5`] = `
<svg
  style={
    Object {
      "height": 300,
      "width": 500,
    }
  }
  viewBox="0 0 500 300"
>
  <path
    d="M 380 201 A 130 130 0 0 0 120 201.00000000000003 "
    fill="rgba(0,0,0,0.35)"
    stroke="rgba(0,0,0,0.8)"
    strokeWidth="4"
  />
  <line
    stroke="black"
    x1={250}
    x2={101.84674891072936}
    y1={200}
    y2={176.53483024396536}
  />
  <text
    fill="rgba(0,0,0,0.7)"
    style={
      Object {
        "fontFamily": "Roboto, sans-serif",
      }
    }
    textAnchor="middle"
    x={82.09298209882661}
    y={173.40614094316072}
  >
    £10
  </text>
  <text
    fill="rgba(0,0,0,0.7)"
    style={
      Object {
        "fontFamily": "Roboto, sans-serif",
      }
    }
    textAnchor="middle"
    x={90}
    y={199.99999999999997}
  >
    £0
  </text>
  <text
    fill="rgba(0,0,0,0.7)"
    style={
      Object {
        "fontFamily": "Roboto, sans-serif",
      }
    }
    textAnchor="middle"
    x={410}
    y={200}
  >
    £200
  </text>
</svg>
`;
